File: problems/CSES/Sorting_and_Searching/maximum_subarray_sum_ii/maximum_subarray_sum_ii.py
import sys
from collections import deque
# sys.setrecursionlimit(200000)

input = sys.stdin.readline

def main():
    n, a, b = map(int, input().split())
    arr = list(map(int, input().split()))

    prefix = [0] * (n + 1)

    for i in range(n):
        prefix[i + 1] = prefix[i] + arr[i]

    queue = deque()
    res = -10**30

    for right in range(a, n + 1):
        new_left = right - a

        while queue and prefix[queue[-1]] >= prefix[new_left]:
            queue.pop()

        queue.append(new_left)

        min_left = right - b

        while queue[0] < min_left:
            queue.popleft()

        res = max(res, prefix[right] - prefix[queue[0]])

    print(res)

main()

File: problems/CSES/Sorting_and_Searching/maximum_subarray_sum_ii/maximum_subarray_sum_ii.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, a, b;
    cin >> n >> a >> b;
    
    vector<int> arr(n);
    vector<ll> prefix(n + 1, 0);

    for (int i = 0; i < n; i++) {
        ll x;
        cin >> x;

        prefix[i + 1] = prefix[i] + x;
    }

    deque<int> q;
    ll res = LLONG_MIN;

    for (int right = a; right < n + 1; right++) {
        int new_left = right - a;

        while (!q.empty() && prefix[q.back()] >= prefix[new_left])
            q.pop_back();
            
        q.push_back(new_left);

        int min_left = right - b;

        while (!q.empty() && q.front() < min_left) {
            q.pop_front();
        }

        res = max(res, prefix[right] - prefix[q.front()]);
    }

    cout << res << '\n';

    return 0;
}
